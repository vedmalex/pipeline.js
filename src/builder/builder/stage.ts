import { Stage } from '../../stage/stage';
import { StageConfig } from '../../stage/StageConfig';
import { UnsetMarker } from '../../stage/types';
import { BuilderDef, StageBuilder } from './../types';


export function stage<TConfig extends StageConfig<any, any>>(
  _def: Partial<BuilderDef<TConfig>> = {}
): StageBuilder<{
  _type: UnsetMarker;
  _input: UnsetMarker;
  _output: UnsetMarker;
  _usage: {
    output: 1;
    run: 1;
    build: 1;
  };
  _run: UnsetMarker;
  _stage: UnsetMarker;
}> {
  return {
    _def: _def as BuilderDef<TConfig>,
    input(input) {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      _def.cfg.input = input;
      return stage({
        ..._def,
        inputs: input,
      }) as any;
    },
    output(output) {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      _def.cfg.output = output;
      return stage({
        ..._def,
        outputs: output,
      }) as any;
    },
    run(run) {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      _def.cfg.run = run as any;
      return stage({
        ..._def,
      }) as any;
    },
    build() {
      if (!_def.cfg) {
        _def.cfg = {} as TConfig;
      }
      return new Stage(_def.cfg) as any;
    },
  };
}
