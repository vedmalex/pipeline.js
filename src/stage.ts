import { CreateError } from './utils/ErrorList'
import {
  AllowedStage,
  CallbackFunction,
  EnsureFunction,
  getStageConfig,
  StageConfig,
  StageRun,
  ValidateFunction,
} from './utils/types'

import { execute_ensure } from './utils/execute_ensure'
import { execute_validate } from './utils/execute_validate'
import { execute_rescue } from './utils/execute_rescue'
import { execute_callback } from './utils/execute_callback'
import { can_fix_error } from './utils/can_fix_error'
import { run_callback_once } from './utils/types'

// make possibility to context be immutable for debug purposes

export class Stage<T = any, C extends StageConfig<T, R> = any, R = T> {
  public get config(): C {
    return this._config
  }
  protected _config!: C
  constructor(config?: AllowedStage<T, C, R>) {
    if (config) {
      let res = getStageConfig(config)
      if (res instanceof Stage) {
        return res
      } else {
        this._config = res
      }
    } else {
      this._config = {} as C
    }
  }

  public get reportName() {
    return `STG:${this._config.name ? this._config.name : ''}`
  }

  public get name() {
    return this._config.name ?? ''
  }

  // может быть вызван как Promise
  // сделать все дубликаты и проверки методов для работы с промисами
  public execute(context: T): Promise<R | T>
  public execute(context: T, callback: CallbackFunction<R | T>): void
  public execute(
    err: Error | undefined,
    context: T,
    callback: CallbackFunction<R | T>,
  ): void
  public execute(
    _err: Error | T | undefined,
    _context?: T | CallbackFunction<R | T>,
    _callback?: CallbackFunction<R | T>,
  ): void | Promise<R | T> {
    // discover arguments
    let err: Error | undefined,
      context: T,
      __callback: CallbackFunction<R | T> | undefined

    if (arguments.length == 1) {
      context = _err as T
      //promise
    } else if (arguments.length == 2) {
      if (typeof _context == 'function') {
        // callback
        context = _err as T
        err = undefined
        __callback = _context as CallbackFunction<R | T>
      } else {
        //promise
        err = _err as Error
        context = _context as T
      }
    } else {
      // callback
      err = _err as Error
      context = _context as T
      __callback = _callback as CallbackFunction<R | T>
    }

    if (!this.run) {
      this.run = this.compile()
    }
    if (!__callback) {
      return new Promise((res, rej) => {
        this.execute(err, context as T, (err?: Error, ctx?: R | T) => {
          if (err) rej(err)
          else res(ctx as R | T)
        })
      })
    } else {
      const callback = run_callback_once(__callback)
      if (err && this._config.run && !can_fix_error(this._config.run)) {
        this.rescue(err, context, callback)
      } else {
        if (this.config.ensure) {
          this.ensure(this.config.ensure, context, (err_?: Error, ctx?: T) => {
            if (err || err_) {
              if (this._config.run && !can_fix_error(this._config.run)) {
                this.rescue(CreateError([err, err_]), ctx ?? context, callback)
              } else {
                // обработка ошибок может происходить внутри функции
                this.run?.(CreateError([err, err_]), ctx ?? context, callback)
              }
            } else {
              this.run?.(undefined, ctx ?? context, callback)
            }
          })
        } else if (this._config.validate) {
          this.validate(
            this._config.validate,
            context,
            (err_?: Error, ctx?: T) => {
              if (err || err_) {
                if (this._config.run && !can_fix_error(this._config.run)) {
                  this.rescue(
                    CreateError([err, err_]),
                    ctx ?? context,
                    callback,
                  )
                } else {
                  // обработка ошибок может происходить внутри функции
                  this.run?.(CreateError([err, err_]), ctx ?? context, callback)
                }
              } else {
                this.run?.(undefined, ctx ?? context, callback)
              }
            },
          )
        } else {
          this.run?.(undefined, context, callback)
        }
      }
    }
  }

  protected stage(
    err: Error | undefined,
    context: T,
    callback: CallbackFunction<T | R>,
  ) {
    if (this._config.run) {
      execute_callback(err, this._config.run, context, (err?: Error) => {
        if (err) {
          this.rescue(err, context, callback)
        } else {
          callback(undefined, context)
        }
      })
    } else {
      this.rescue(
        CreateError([this.reportName + ' reports: run is not a function', err]),
        context,
        callback,
      )
    }
  }

  public compile(rebuild: boolean = false): StageRun<T, R> {
    let res: StageRun<T, R>
    if (this.config.precompile) {
      this.config.precompile()
    }
    if (this._config.compile) {
      res = this._config.compile.call(this, rebuild)
    } else if (!this.run || rebuild) {
      res = this.stage
    } else {
      res = this.run
    }
    return res
  }
  // to be overridden by compile
  protected run?: StageRun<T, R>

  // объединение ошибок сделать
  protected rescue(
    err: Error | undefined,
    context: T,
    callback: CallbackFunction<T | R>,
  ) {
    if (err && !(err instanceof Error)) {
      if (typeof err == 'string') err = CreateError(err)
    }
    if (err && this._config.rescue) {
      execute_rescue(this._config.rescue, err, context, (_err?: Error) => {
        callback(_err, context)
      })
    } else {
      // отправить ошибку дальше
      callback(err, context)
    }
  }

  public toString() {
    return '[pipeline Stage]'
  }

  protected validate(
    validate: ValidateFunction<T>,
    context: T,
    callback: CallbackFunction<T>,
  ) {
    execute_validate(
      validate,
      context,
      (err: Error | undefined, result: boolean | undefined) => {
        if (err) {
          callback(err, context)
        } else {
          if (result) {
            if ('boolean' === typeof result) {
              callback(undefined, context)
            } else if (Array.isArray(result)) {
              callback(CreateError(result))
            }
          } else {
            callback(CreateError(this.reportName + ' reports: T is invalid'))
          }
        }
      },
    )
  }
  protected ensure(
    ensure: EnsureFunction<T>,
    context: T,
    callback: CallbackFunction<T>,
  ) {
    execute_ensure(
      ensure,
      context,
      (err: Error | undefined, result: T | undefined) => {
        callback(err, result ?? context)
      },
    )
  }
}

export type EnsureParams<T, R> = {
  context: T
  callback: CallbackFunction<R> | undefined
  err: Error | undefined
  is_promise: boolean
}
