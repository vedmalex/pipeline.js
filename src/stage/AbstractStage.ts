import { z } from 'zod'

import { fromZodError } from 'zod-validation-error'
import { BaseStageConfig, Run, RunConfig, validatorRun } from './StageConfig'
import { makeCallbackArgs } from './types'

export class AbstractStage<
  Input,
  Output,
  TConfig extends BaseStageConfig<Input, Output> = BaseStageConfig<Input, Output>,
> {
  private _config: BaseStageConfig<Input, Output> & RunConfig<Input, Output>
  get config(): TConfig {
    return this._config as unknown as TConfig
  }
  protected set config(config: TConfig) {
    this._config = config as any
  }
  protected get run() {
    return this._config.run
  }
  protected set run(run: Run<Input, Output>) {
    this._config.run = run
  }
  constructor(config: BaseStageConfig<Input, Output> & RunConfig<Input, Output>) {
    if (typeof config === 'object' && config !== null) {
      this._config = config
      this._config.run = this._config.run.bind(this)
    } else {
      throw new Error('arguments Error')
    }
    this.exec = this.exec.bind(this)
  }

  // может быть вызван как Promise
  // сделать все дубликаты и проверки методов для работы с промисами
  public async exec(input: Input): Promise<Output> {
    if (this._config.input) {
      const validateResult = await this.validate(this._config.input, input)
      if (validateResult.result === 'failure') {
        throw validateResult.reason
      }
    }

    const result = await validatorRun(this.config.input, this.config.output).parse(this._config.run)(input)

    if (this._config.output) {
      const validateResult = await this.validate(this._config.output, result)
      if (validateResult.result === 'failure') {
        throw validateResult.reason
      }
    }
    return result
  }

  protected async validate<T>(
    validate: z.ZodType<T>,
    context: unknown,
  ) {
    const result = await validate.safeParseAsync(context)
    if (!result.success) {
      return makeCallbackArgs(fromZodError(result?.error), context as T)
    } else {
      return makeCallbackArgs(undefined, context as T)
    }
  }
}
