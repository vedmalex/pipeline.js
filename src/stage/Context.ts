import { defaultsDeep, get, set } from 'lodash'

import { StageObject } from './types'

export const ContextSymbol = Symbol('Context')
export const OriginalObject = Symbol('OriginalObject')
export const ProxySymbol = Symbol('Handler')
export interface ContextProxy<T> {
  fork<Config extends T>(config: Partial<Config>): ProxyType<T & Config>
  // addChild(child: object): unknown
  get(path: keyof T): any
  // addSubtree(lctx: object): unknown
  getParent(): ProxyType<T>
  getRoot(): ProxyType<T>
  setParent(parent: ProxyType<T>): void
  setRoot(parent: ProxyType<T>): void
  hasChild(ctx: object): boolean
  // hasSubtree(ctx: object): boolean
  toObject(clean?: boolean): T
  toJSON(): string
  toString(): string
  get original(): T
  [OriginalObject]?: true
  [key: string | symbol | number]: any
}

export type ProxyType<T> = ContextProxy<T> & T
export type ContextType<T> = T extends StageObject ? ProxyType<T> : T

var count = 0
var allContexts: Record<string, Context<any>> = {}

export enum RESERVATIONS {
  prop,
  func_this,
  func_ctx,
}

export const RESERVED = {
  getParent: RESERVATIONS.func_ctx,
  getRoot: RESERVATIONS.func_ctx,
  setParent: RESERVATIONS.func_ctx,
  setRoot: RESERVATIONS.func_ctx,
  toString: RESERVATIONS.func_ctx,
  original: RESERVATIONS.prop,
  __parent: RESERVATIONS.prop,
  __root: RESERVATIONS.prop,
  __stack: RESERVATIONS.prop,
  hasChild: RESERVATIONS.func_ctx,
  hasSubtree: RESERVATIONS.func_ctx,
  ensure: RESERVATIONS.func_ctx,
  addChild: RESERVATIONS.func_ctx,
  addSubtree: RESERVATIONS.func_ctx,
  toJSON: RESERVATIONS.func_ctx,
  toObject: RESERVATIONS.func_ctx,
  fork: RESERVATIONS.func_this,
  get: RESERVATIONS.func_this,
  allContexts: RESERVATIONS.func_this,
}
/**
 *  The **Context** itself
 *  @param {Object} config The object that is the source for the **Context**.
 */
export class Context<Input extends StageObject> implements ContextProxy<Input> {
  public static ensure<Input>(_config?: Input): ProxyType<Input> {
    if (Context.isProxy<Input>(_config)) {
      return _config
    } else {
      if (typeof _config === 'object' && _config !== null) {
        return this.create(_config)
      } else {
        return this.create()
      }
    }
  }

  public static create<Input>(input?: unknown): ProxyType<Input> {
    return new Context(input ?? {}) as unknown as ProxyType<Input>
  }

  public static isProxy<Input>(obj?: unknown): obj is ProxyType<Input> {
    return typeof obj == 'object' && obj !== null ? obj[ContextSymbol] : false
  }

  protected ctx: Input extends StageObject ? Input : never
  protected proxy: any
  protected __parent!: ProxyType<Input>
  protected __root!: ProxyType<Input>
  protected __stack?: string[]
  protected id: number;
  [OriginalObject]?: true
  get original() {
    return this.ctx
  }

  protected constructor(config: Input extends StageObject ? Input : never) {
    this.ctx = config
    this.id = count++
    allContexts[this.id] = this
    const res = new Proxy(this, {
      get(target: Context<Input>, key: string | symbol | number, _proxy: any): any {
        if (key == ContextSymbol) {
          return true
        }
        if (key == ProxySymbol) {
          return _proxy
        }
        if (key == 'allContexts') {
          return allContexts
        }

        if (!(key in RESERVED)) {
          if (key in target.ctx) {
            return target.ctx[key]
          } else {
            return target.__parent?.[key]
          }
        } else {
          if (RESERVED[key as keyof typeof RESERVED] == RESERVATIONS.func_ctx) {
            return target[key].bind(target)
          }
          if (RESERVED[key as keyof typeof RESERVED] == RESERVATIONS.func_this) {
            return target[key].bind(target)
          } else {
            return target[key] // just props
          }
        }
      },
      set(target: Context<Input>, key: keyof typeof RESERVED | string | symbol | number, value): boolean {
        if (!(key in RESERVED)) {
          target.ctx[key] = value
          return true
        } else if (
          typeof key == 'string'
          && key in RESERVED
          && RESERVED[key as keyof typeof RESERVED] != RESERVATIONS.prop
        ) {
          return false
        } else {
          target[key] = value
          return true
        }
      },
      deleteProperty(target: Context<Input>, key: string | symbol) {
        if (!(key in RESERVED)) {
          return delete target.ctx[key]
        } else {
          return false
        }
      },
      has(target: Context<Input>, key: string | symbol) {
        if (!(key in RESERVED)) {
          if (target.__parent) {
            return key in target.ctx || key in target.__parent
          } else {
            return key in target.ctx
          }
        } else {
          return false
        }
      },
      ownKeys(target: Context<Input>) {
        if (target.__parent) {
          return [...Reflect.ownKeys(target.ctx), ...Reflect.ownKeys(target.__parent)]
        } else {
          return Reflect.ownKeys(target.ctx)
        }
      },
    })

    this.proxy = res

    return res
  }

  /**
   * Makes fork of current context and add it to current as a child context
   * @api public
   * @param {Object|Context} [config] new properties that must exists in new fork
   * @retrun {Context}
   */
  fork<Child extends Input>(ctx?: Partial<Child>): ProxyType<Input & Child> {
    var child = Context.ensure<Child>(ctx as Child)
    this.addChild(child)
    return child as ProxyType<Input & Child>
  }

  protected addChild<Child extends Input>(child: ProxyType<Child>) {
    if (!this.hasChild(child)) {
      child.setParent(this.proxy)
    }
  }

  /**
   * Same but different as a fork. it make possible get piece of context as context;
   * @param path String path to context object that need to be a Context instance
   * @return {Context} | {Primitive type}
   */
  get(path: keyof Input): any {
    var root = get(this.ctx, path) as any
    if (root instanceof Object) {
      var result = root
      if (!Context.isProxy(result)) {
        var lctx = Context.ensure(result)
        this.addSubtree(lctx)
        set(this.original, path, lctx)
        result = lctx
      }
      return result
    } else {
      return root
    }
  }

  protected addSubtree(lctx: object): unknown {
    if (Context.isProxy<Input>(lctx)) {
      if (!this.hasSubtree(lctx)) {
        lctx.setRoot(this.proxy)
      }
      return lctx
    } else {
      return lctx
    }
  }

  /**
   * Return parent Context
   * @api public
   * @return {Context}
   */
  getParent() {
    return this.__parent
  }
  getRoot() {
    return this.__root
  }
  setParent(parent: ProxyType<Input>) {
    this.__parent = parent
  }
  setRoot(root: ProxyType<Input>) {
    this.__root = root
  }
  /**
   * checks wheater or not context has specific child context
   * it return `true` also if `ctx` is `this`;
   * @api public
   * @return {Boolean}
   */
  hasChild(ctx: unknown): boolean {
    if (Context.isProxy<Input>(ctx) && ctx.__parent) {
      return ctx.__parent == this.proxy || this.proxy == ctx
    } else {
      return false
    }
  }
  protected hasSubtree(ctx: unknown): boolean {
    if (Context.isProxy<Input>(ctx) && ctx.__root) {
      return ctx.__root == this.proxy || this.proxy == ctx
    } else {
      return false
    }
  }

  /**
   * Convert context to raw Object;
   * @api public
   * @param {Boolean} [clean]  `true` it need to clean object from referenced Types except Function and raw Object(js hash)
   * @return {Object}
   */
  toObject(): Input {
    const obj = {} as Input
    defaultsDeep(obj, this.ctx)
    if (this.__parent) {
      // TODO: взять весь объект по всей структуре дерева
      defaultsDeep(obj, this.__parent.toObject(false))
    }
    return obj
  }

  /**
   * Conterts context to JSON
   * @api public
   * @return {String}
   */
  toJSON(): string {
    // always cleaning the object
    return JSON.stringify(this.toObject())
  }

  /**
   * toString
   */
  toString(): string {
    return '[pipeline Context]'
  }
}
