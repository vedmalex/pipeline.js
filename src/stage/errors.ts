export const ERROR = {
  signature: 'unacceptable run method signature',
  invalid_context: 'context is invalid',
}
