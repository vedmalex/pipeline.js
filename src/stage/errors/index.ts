export * from './can_fix_error'
export * from './CreateError'
export * from './ErrorList'
export * from './errors'
export * from './isComplexError'
export * from './process_error'
