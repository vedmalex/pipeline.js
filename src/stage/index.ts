export * from './Context'
export * from './errors'
export * from './getStageConfig'
export * from './stage'
export * from './StageConfig'
export * from './types'
export * from './utils'
