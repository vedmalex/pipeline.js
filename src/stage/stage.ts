import { z } from 'zod'

import { fromZodError } from 'zod-validation-error'
import { Context, OriginalObject } from './Context'
import { can_fix_error, ComplexError, CreateError } from './errors'
import { getStageConfig, isAnyStage, StageSymbol } from './getStageConfig'
import { StageConfig } from './StageConfig'
import { AllowedStage, AnyStage, CallbackFunction, isStageRun, Possible, RunPipelineFunction, StageRun } from './types'
import { execute_callback } from './utils/execute_callback'
import { execute_custom_run } from './utils/execute_custom_run'
import { execute_rescue } from './utils/execute_rescue'

export class Stage<R, C extends StageConfig<R> = StageConfig<R>> implements AnyStage<R> {
  public get config(): C {
    return this._config as C
  }
  [StageSymbol]: boolean
  protected _config!: C
  constructor()
  constructor(name: string)
  constructor(config: C)
  constructor(runFn: RunPipelineFunction<R>)
  constructor(stage: AnyStage<R>)
  constructor(config?: AllowedStage<R, C>) {
    this[StageSymbol] = true
    if (config) {
      if (typeof config === 'string') {
        this._config = { name: config } as C extends StageConfig<R> ? C : never
      } else {
        let res = getStageConfig<R, C>(config) as C extends StageConfig<R> ? C : never
        if (isAnyStage<R>(res)) {
          this._config = res.config as any
        } else {
          this._config = res
        }
      }
    } else {
      this._config = {} as C extends StageConfig<R> ? C : never
    }
  }

  public get reportName() {
    return `STG:${this._config.name ? this._config.name : ''}`
  }

  public toString() {
    return '[pipeline Stage]'
  }

  public get name() {
    return this._config.name ?? ''
  }

  // может быть вызван как Promise
  // сделать все дубликаты и проверки методов для работы с промисами
  public execute(context: R): Promise<R>
  public execute(context: R, callback: CallbackFunction<R>): void
  public execute(err: unknown, context: R, callback: CallbackFunction<R>): void
  public execute(
    _err?: unknown,
    _context?: unknown,
    _callback?: unknown,
  ): void | Promise<R> {
    // discover arguments
    let err: Possible<ComplexError>
    let not_ensured_context: R | R
    let __callback: Possible<CallbackFunction<R>> = undefined

    if (arguments.length == 1) {
      not_ensured_context = _err as R | R
      // promise
    } else if (arguments.length == 2) {
      if (typeof _context == 'function') {
        // callback
        not_ensured_context = _err as R | R
        err = undefined
        __callback = _context as CallbackFunction<R>
      } else {
        // promise
        err = _err as ComplexError
        not_ensured_context = _context as R | R
      }
    } else {
      // callback
      err = _err as ComplexError
      not_ensured_context = _context as R | R
      __callback = _callback as CallbackFunction<R>
    }

    if (!this.run) {
      this.run = this.compile()
    } else if (!this.config?.run) {
      // legacy run
      if (!isStageRun(this.run)) {
        var legacy = this.run

        this.run = execute_custom_run<R>(legacy as RunPipelineFunction<R>)
      }
    }

    const stageToRun = this.run?.bind(this)

    const input_is_context = Context.isProxy(not_ensured_context)
    const isProxible = input_is_context || (not_ensured_context && typeof not_ensured_context === 'object')

    let context = isProxible ? Context.ensure(not_ensured_context) : not_ensured_context
    if (isProxible) {
      context[OriginalObject] = true
    }
    // выполнить валидацию результата сразу же перед вызовом последнего значения
    if (!__callback) {
      return new Promise<R>((res, rej) => {
        this.execute(err, context, (err, ctx) => {
          if (err) {
            rej(err)
          } else {
            if (input_is_context) {
              res(ctx ?? context)
            } else {
              if (Context.isProxy(ctx)) {
                res(ctx.original as R)
              } else {
                res(ctx as unknown as R)
              }
            }
          }
        })
      })
    } else {
      const back: typeof __callback = (err, _ctx) => {
        if (input_is_context) {
          __callback?.(err, _ctx)
        } else {
          if (Context.isProxy<R>(_ctx)) {
            __callback?.(err, _ctx.original as R)
          } else {
            __callback?.(err, _ctx)
          }
        }
      }
      const successRescue = (ret: R) => {
        if (this._config.output) {
          this.validate(this._config.output, ret ?? context, (err_, ctx) => {
            if (err_) {
              this.rescue(err_, ctx ?? context, fail, successRescue)
            } else {
              back(err, ctx ?? context)
            }
          })
        } else {
          success(ret)
        }
      }
      const success = (ret: R) => back(undefined, ret ?? context)
      const fail = (err: unknown) => back(err, context)

      const callback = (err?, _ctx?: R) => {
        if (err) {
          this.rescue(err, _ctx ?? context, fail, successRescue)
        } else {
          if (this._config.output) {
            this.validate(this._config.output, _ctx ?? context, (err_, ctx) => {
              if (err_) {
                this.rescue(err_, ctx ?? context, fail, successRescue)
              } else {
                back(err, ctx ?? context)
              }
            })
          } else {
            back(err, _ctx ?? context)
          }
        }
      }

      if (err && this._config.run && !can_fix_error(this._config.run)) {
        this.rescue(err, context, fail, successRescue)
      } else {
        if (this._config.input) {
          this.validate(this._config.input, context, (err_, ctx) => {
            this.runStageMethod(err, err_, ctx, context, stageToRun, callback)
          })
        } else {
          stageToRun(undefined, context, callback)
        }
      }
    }
  }

  protected runStageMethod(
    err_: unknown,
    err: unknown,
    ctx: R | undefined,
    context: R,
    stageToRun: StageRun<R>,
    callback: CallbackFunction<R>,
  ) {
    if (err || err_) {
      if (this.config.run && !can_fix_error(this.config.run)) {
        this.rescue(CreateError([err, err_] as Array<Error>), ctx ?? context, callback, rescuedContext => {
          // ошибка обработана все хорошо, продолжаем
          stageToRun(undefined, rescuedContext, callback)
        })
      } else {
        // обработка ошибок может происходить внутри функции
        stageToRun(CreateError([err, err_] as Array<Error>), ctx ?? context, callback)
      }
    } else {
      stageToRun(undefined, ctx ?? context, callback)
    }
  }

  protected stage(err: unknown, context: R, callback: CallbackFunction<R>) {
    const back = callback
    const sucess = (ret: unknown) => back(undefined, (ret ?? context) as R)
    const fail = (err: unknown) => back(err, context)
    if (this._config.run) {
      if (context) {
        ;(execute_callback<R>).call(this, err, this._config.run, context, callback)
      } else {
        // возвращаем управление
        callback(err)
      }
    } else {
      const retErr: Array<any> = [this.reportName + ' reports: run is not a function']
      if (err) {
        retErr.push(err)
      }
      this.rescue(CreateError(retErr), context, fail, sucess)
    }
  }

  // to be overridden by compile
  protected run?: StageRun<R>
  /**
   * Compile the pipeline.
   * @param rebuild Whether to rebuild the pipeline.
   * @returns The compiled pipeline.
   */
  protected compile(rebuild: boolean = false): StageRun<R> {
    let res: StageRun<R>
    if (this.config.compile) {
      res = this.config.compile.call(this, rebuild)
    } else if (!this.run || rebuild) {
      res = this.stage
    } else {
      if (isStageRun<R>(this.run)) {
        res = this.run
      } else {
        res = execute_custom_run(this.run)
      }
    }
    return res
  }

  // объединение ошибок сделать
  // посмотреть что нужно сделать чтобы вызвать ошибку правильно!!!
  // в начале выполнения важен правильный callback, возможно без контекста
  // в конце важен и контекст ошибки? или не важен
  protected rescue(
    _err: unknown,
    context: R,
    fail: (err: unknown) => void,
    success: (ctx: R) => void,
  ) {
    let err: Possible<ComplexError>

    if (_err) {
      if (!(_err instanceof Error)) {
        err = CreateError(_err)
      } else {
        err = CreateError(_err)
      }
    } else {
      err = null
    }

    if (err && this._config.rescue) {
      execute_rescue(this._config.rescue, err, context, _err => {
        // здесь может быть исправлена ошибка, и контекст передается дальше на выполнение
        if (_err) {
          fail(_err)
        } else {
          success(context)
        }
      })
    } else {
      // отправить ошибку дальше
      // окончателная ошибка и выходим из выполнения
      if (err) {
        fail(err)
      } else {
        success(context)
      }
    }
  }

  protected rescue_async(
    _err: unknown,
    context: R,
  ): Promise<[unknown, R]> {
    return new Promise(resolve => {
      this.rescue(_err, context, err => {
        resolve([err, context])
      }, res => {
        resolve([undefined, res ?? context])
      })
    })
  }

  protected validate(
    validate: z.ZodType<R>,
    context: R,
    callback: CallbackFunction<R>,
  ) {
    validate.safeParseAsync(context)
      .then(result => {
        if (!result.success) {
          callback(CreateError(fromZodError(result?.error)), context)
        } else {
          callback(undefined, context)
        }
      })
  }
}
