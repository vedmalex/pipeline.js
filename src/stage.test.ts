import { JSONSchemaType } from 'ajv'
import 'jest'
import { Stage } from './stage'

describe('stage', () => {
  it('throw error', done => {
    let st = new Stage()
    expect.assertions(1)
    st.execute({}, (err, context) => {
      expect(err).not.toBeUndefined()
      done()
    })
  })

  it('empty run', done => {
    new Stage().execute({}, (err, res) => {
      expect(err).not.toBeUndefined()
      expect(res).toMatchObject({})
      done()
    })
  })
  it('create named', () => {
    const s = new Stage('name')
    expect(s).not.toBeNull()
    expect(s).toMatchSnapshot('names stage')
  })
  it('create with function', () => {
    const s = new Stage<{ name?: string }>(function RunStage(this: { name?: string }) {
      this.name = 'run this Stage'
    })
    expect(s).not.toBeNull()
    expect(s).toMatchSnapshot('function stage')
  })

  it('create with Lambda 3', () => {
    const s = new Stage<{ name: string }>((err, ctx, done) => {
      if (!err) {
        ctx.name = 'run the stage'
        done(undefined, ctx)
      } else {
        done(err)
      }
    })
    expect(s).toMatchSnapshot('lambda stage')
  })

  it('creates throws when both parameters validate and schema are passed', () => {
    expect(
      () =>
        new Stage({
          run: () => {},
          schema: {},
          validate: {},
        } as any),
    ).toThrow()
  })

  it('intialize using schema and validate separately', () => {
    expect(
      () =>
        new Stage<{ name?: string }>({
          run: () => {},
          validate: (_ctx: {}) => true,
        }),
    ).not.toThrow()
    expect(
      () =>
        new Stage({
          run: () => {},
          schema: {} as JSONSchemaType<{}>,
        }),
    ).not.toThrow()
  })

  it('validate using schema', () => {
    const st = new Stage<{ name: string }>({
      run: () => {},
      schema: {
        type: 'object',
        properties: {
          name: { type: 'string' },
        },
        required: ['name'],
        additionalProperties: false,
        errorMessage: 'should be an object with a name property',
      },
    })

    st.execute({ fullname: 1 } as unknown as { name: string }, (err, res) => {
      expect(err).not.toBeUndefined()
      expect(err).toMatchSnapshot()
    })
  })

  it('initialize other stuff sucessfully', () => {
    let stage = new Stage({
      run: () => {},
      ensure: () => {},
      rescue: () => {},
      name: 'stage',
    })
    expect(stage.name).toBe('stage')
    expect(stage.reportName).toBe(`STG:${stage.name}`)
    expect(stage).toMatchSnapshot('schema stage 1')
  })

  it('create with Lambda 2', () => {
    const s = new Stage((err, ctx) => {
      if (!err) {
        ctx.name = 'run the stage'
        return Promise.resolve(ctx)
      } else {
        return Promise.reject(err)
      }
    })
    expect(s).toMatchSnapshot('lambda stage')
  })

  it('create with Config', () => {
    const s = new Stage<{ name: string }>((err, ctx, done) => {
      if (!err) {
        ctx.name = 'run the stage'
        done(undefined, ctx)
      } else {
        done(err)
      }
    })
    expect(s).toMatchSnapshot('config stage')
  })

  // it('use Precompile to run everything', () => {
  //   type CTX = {
  //     name: string
  //   }
  //   interface Config extends StageConfig<CTX, CTX> {
  //     stages: Array<IStage<any, any> | RunPipelineFunction>
  //     addStage(
  //       this: Config,
  //       stage:
  //         | StageConfig<CTX, CTX>
  //         | SingleStageFunction
  //         | IStage<any, any>,
  //     ): void
  //   }

  //   const pipeline = new Stage<CTX, CTX, Config>({
  //     stages: [],
  //     precompile(this: Config) {
  //       let run: StageRun<CTX, CTX> = (
  //         err: Possible<ComplexError>,
  //         context: CTX,
  //         done: CallbackFunction,
  //       ) => {
  //         let i = -1
  //         //sequential run;
  //         let next = (err: Possible<ComplexError>, context: CTX | undefiend) => {
  //           i += 1
  //           if (i < this.stages.length) {
  //             run_or_execute(this.stages[i], err, context, next)
  //           } else if (i == this.stages.length) {
  //             done(err, context)
  //           } else {
  //             done(new Error('done call more than once'), context)
  //           }
  //         }
  //         next(err, context)
  //       }

  //       if (this.stages.length > 0) {
  //         this.run = run
  //       } else {
  //         this.run = empty_run
  //       }
  //     },
  //     addStage(
  //       _stage:
  //         | StageConfig<CTX, CTX>
  //         | RunPipelineFunction
  //         | IStage<any, any>,
  //     ) {
  //       let stage: IStage<any, any> | RunPipelineFunction| undefined
  //       if (typeof _stage === 'function') {
  //         stage = _stage
  //       } else {
  //         if (typeof _stage === 'object') {
  //           if (!isIStage(_stage)) {
  //             stage = new Stage(_stage)
  //           } else {
  //             stage = _stage
  //           }
  //         }
  //       }
  //       if (stage) {
  //         this.stages.push(stage)
  //         this.run = undefined
  //       }
  //     },
  //   })
  // })
})
