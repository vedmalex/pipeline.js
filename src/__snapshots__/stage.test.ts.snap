// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stage create named: names stage 1`] = `
Stage {
  "_config": {
    "name": "name",
  },
  Symbol(stage): true,
}
`;

exports[`stage create with Config: config stage 1`] = `
Stage {
  "_config": {
    "name": "(err, ctx, done) => {
            if (!err) {
                ctx.name = 'run the stage';
                done(undefined, ctx);
            }
            else {
                done(err);
            }
        }",
    "run": [Function],
  },
  Symbol(stage): true,
}
`;

exports[`stage create with Lambda 2: lambda stage 1`] = `
Stage {
  "_config": {
    "name": "(err, ctx) => {
            if (!err) {
                ctx.name = 'run the stage';
                return Promise.resolve(ctx);
            }
            else {
                return Promise.reject(err);
            }
        }",
    "run": [Function],
  },
  Symbol(stage): true,
}
`;

exports[`stage create with Lambda 3: lambda stage 1`] = `
Stage {
  "_config": {
    "name": "(err, ctx, done) => {
            if (!err) {
                ctx.name = 'run the stage';
                done(undefined, ctx);
            }
            else {
                done(err);
            }
        }",
    "run": [Function],
  },
  Symbol(stage): true,
}
`;

exports[`stage create with function: function stage 1`] = `
Stage {
  "_config": {
    "name": "RunStage",
    "run": [Function],
  },
  Symbol(stage): true,
}
`;

exports[`stage initialize other stuff sucessfully: schema stage 1 1`] = `
Stage {
  "_config": {
    "name": "stage",
    "run": [Function],
  },
  Symbol(stage): true,
}
`;
